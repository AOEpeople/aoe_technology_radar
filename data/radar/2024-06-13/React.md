---
title: "React"
ring: adopt
quadrant: languages-and-frameworks
featured: true
tags: [Framework, Frontend, Web Development]
---

[React](https://reactjs.org/) is a JavaScript library for building user interfaces. It allows developers to create large web applications that can update and render efficiently in response to data changes.

MOHARA uses React for building dynamic and responsive web applications. Its component-based architecture and virtual DOM provide excellent performance and maintainability.
