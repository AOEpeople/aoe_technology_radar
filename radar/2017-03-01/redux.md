---
title:      "Redux"
ring:       productize
quadrant:   data-science-and-analytics

---

[Redux](http://redux.js.org/) helps us to maintain state in our frontend applications in a more predictable and clearer way. It is extendable though middleware, it has a great documentation and some awesome [devui-and-devices](https://github.com/gaearon/redux-devui-and-devices) that are especially helpful when you are new to Redux.

The functional concepts for updating the state, combined with immutable data, lead to extremely easy and enjoyable [unit tests](http://redux.js.org/docs/recipes/WritingTests.html) - this is maybe the biggest plus for us developers.

The official [react-redux bindings](https://github.com/reactjs/react-redux) also made it straightforward to weave Redux into our React applications. For asynchronous actions we use [redux-sagas](https://redux-saga.github.io/redux-saga/) which has proven itself as a better alternative for [redux-thunk](https://github.com/gaearon/redux-thunk).

Currently, we use Redux only in our React projects, but we are evaluating it together with other frameworks such as Angular or Vue.js, as well.
