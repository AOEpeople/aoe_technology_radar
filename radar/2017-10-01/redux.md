---
title:      "Redux"
ring:       trial
quadrant:   languages-and-frameworks

---

[Redux](http://redux.js.org/) helps us to maintain state in our frontend applications in a more predictable and clearer way. It is extendable though middleware, it has a great documentation and some awesome [devtools](https://github.com/gaearon/redux-devtools) that are especially helpful when you are new to Redux.

The functional concepts for updating the state, combined with immutable data, lead to extremely easy and enjoyable [unit tests](http://redux.js.org/docs/recipes/WritingTests.html) - this is maybe the biggest plus for us developers.

The official [react-redux bindings](https://github.com/reactjs/react-redux) also made it straightforward to weave Redux into our React applications. For asynchronous actions we use [redux-sagas](https://redux-saga.github.io/redux-saga/) which has proven itself as a better alternative for [redux-thunk](https://github.com/gaearon/redux-thunk).

Currently, we use Redux only in our React projects, but we are evaluating it together with other frameworks such as Angular or Vue.js, as well.
