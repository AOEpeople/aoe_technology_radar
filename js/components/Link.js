import React from 'react';
import { connect } from 'react-redux';
import actions from '../actions';

function Link({ pageName, children, navigate, className, style = {}}) {
  const handleClick = (e) => {
    e.preventDefault();
    navigate(pageName);
  };

  return (
    <a href={`/techradar/${pageName}.html`} onClick={handleClick} style={style} {...{ className }}>
      {children}
    </a>
  );
}

export default connect(
  undefined,
  actions
)(Link);
