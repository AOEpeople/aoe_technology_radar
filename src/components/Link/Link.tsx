import React from "react";
import { Link as RLink } from "react-router-dom";

import { useSearchParamState } from "../../hooks/use-search-param-state";
import "./link.scss";

type LinkProps = {
  pageName: string;
  style?: React.CSSProperties;
  className?: string;
};

function Link({
  pageName,
  children,
  className,
  style = {},
}: React.PropsWithChildren<LinkProps>) {
  const [searchParamState] = useSearchParamState(undefined, {
    parseOptions: { decode: false },
  });
  const { tags } = searchParamState;

  return (
    <RLink
      to={tags ? `/${pageName}.html?tags=${tags}` : `/${pageName}.html`}
      style={style}
      {...{ className }}
    >
      {children}
    </RLink>
  );
}

export default Link;
