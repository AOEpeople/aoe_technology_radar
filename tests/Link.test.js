import React from 'react';
import { Provider, connect } from "react-redux";
import renderer from 'react-test-renderer';
import { createStore, applyMiddleware } from 'redux';

import appReducer from '../js/reducer';
import Link from '../js/components/Link.js';

test('Renders a Link component', () => {
  const preloadedState = window.__TECHRADAR__;
  const store = createStore(
    appReducer,
    preloadedState,
  );
  const component = renderer.create(
    <Provider store={store}>
      <Link pageName="overview" className="icon-link">
        <span className="icon icon--overview icon-link__icon"></span>Technologies Overview
      </Link>
    </Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});