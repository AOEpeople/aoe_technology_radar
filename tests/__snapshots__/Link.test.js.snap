// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a Link component 1`] = `
<a
  className="icon-link"
  href="/techradar/overview.html"
  onClick={[Function]}
  style={Object {}}
>
  <span
    className="icon icon--overview icon-link__icon"
  />
  Technologies Overview
</a>
`;
